src/saveload/marker.rs

```diff
@@ -0,0 +1,209 @@
+///     id: u64,
+///     seq: u64,
+/// }
+///
+/// impl Component for NetMarker {
+///     type Storage = DenseVecStorage<Self>;
+/// }
+///
+/// impl Marker for NetMarker {
+///     type Identifier = u64;
+///     type Allocator = NetNode;
+///
+///     fn id(&self) -> u64 {
+///         self.id
+///     }
+///
+///     // Updates sequence id.
+///     // Entities with too old sequence id get deleted.
+///     fn update(&mut self, update: Self) {
+///         assert_eq!(self.id, update.id);
+///         self.seq = update.seq;
+///     }
+/// }
+///
+/// // Each client and server has one
+/// // Contains id range and `NetMarker -> Entity` mapping
+/// struct NetNode {
+///     range: Range<u64>,
+///     mapping: HashMap<u64, Entity>,
+/// }
+///
+/// impl MarkerAllocator<NetMarker> for NetNode {
+///     fn allocate(&mut self, entity: Entity, id: Option<u64>) -> NetMarker {
+///         let id = id.unwrap_or_else(|| {
+///             self.range.next().expect("Id range must be virtually endless")
+///         });
+///         let marker = NetMarker {
+///             id: id,
+///             seq: 0,
+///         };
+///         self.mapping.insert(id, entity);
+///         marker
+///     }
+///
+///     fn get(&self, id: u64) -> Option<Entity> {
+///         self.mapping.get(&id).cloned()
+///     }
+/// }
+///
+/// fn main() {}
+/// ```
+pub trait Marker: SerializableComponent + Copy {
+    /// Id of the marker
+    type Identifier: Copy + ::std::fmt::Debug + Eq + Hash;
+
+    /// Allocator for this `Marker`
+    type Allocator: MarkerAllocator<Self>;
+
+    /// Get this marker internal id
+    fn id(&self) -> Self::Identifier;
+
+    /// Update marker with new value.
+    /// It must preserve internal `Identifier`.
+    ///
+    /// # Panics
+    ///
+    /// Allowed to panic if `self.id() != update.id()`.
+    /// But usually implementer may ignore `update.id()` value
+    /// as deserialization algorithm ensures `id()`'s match.
+    fn update(&mut self, update: Self) {
+        ::std::mem::drop(update);
+    }
+}
+
+/// This allocator is used with `Marker` trait
+/// It provides method for allocation of `Marker`s
+/// And also should provide `Marker -> Entity` mapping
+/// `maintain` method can be implemented for cleanup and actualization
+/// # Example
+/// see docs for `Marker`
+pub trait MarkerAllocator<M: Marker>: Resource {
+    /// Allocate new `Marker`.
+    /// Stores mapping `Marker` -> `Entity`.
+    fn allocate(&mut self, entity: Entity, id: Option<M::Identifier>) -> M;
+
+    /// Get `Entity` by `Marker::Identifier`
+    fn get(&self, id: M::Identifier) -> Option<Entity>;
+
+    /// Create new unique marker `M` and attach it to entity.
+    /// Or get old marker if already marked.
+    fn mark<'a>(&mut self, entity: Entity, storage: &mut WriteStorage<'a, M>) -> M {
+        if let Some(marker) = storage.get(entity).cloned() {
+            marker
+        } else {
+            let marker = self.allocate(entity, None);
```

Comment:
Should we just use `None` here?

Reply:
The `None` means allocator should come up with new id.

Reply:
Probably it must be documented

Reply:
This `None` asks to allocate new `Marker` with new unique `Identifier`.
If one has `Identifier` already but need new `Marker` he provides it in this function.

Reply:
I'll add comment to `allocate` method about `id` argument.